### PR TITLE
[InstCombine] Implement `fcmp (fadd x, 0.0), y` => `fcmp x, y` optimization

### PR DESCRIPTION
**Pull request by vmishelcs**
This PR addresses issue #88168. I have implemented an optimization for the case of

```
define i1 @fcmp_fadd_zero_ugt(float %x, float %y) {
  %add = fadd float %x, 0.000000e+00
  %cmp = fcmp ugt float %add, %y
  ret i1 %cmp
}
```
`=>`
```
define i1 @fcmp_fadd_zero_ugt(float %x, float %y) {
  %cmp = fcmp ugt float %x, %y
  ret i1 %cmp
}
```
and all other types of `fcmp` instructions (`uge`, `ogt`, etc). I have also added tests to `llvm/test/Transforms/InstCombine/fcmp.ll` to check this behaviour.

Proofs: 
`fadd x, 0.0` https://alive2.llvm.org/ce/z/7FzNnM
`fsub x, 0.0` https://alive2.llvm.org/ce/z/puUxLK

### CODE REVIEW

**Comment by vmishelcs** [comment]
@dtcxzyw Please take a look whenever you have time :) Let me know if I should make any changes.

Question: Should I also add tests for the case of `fcmp (fsub x, 0.0), y`? From what I can tell my implementation already covers this case. Is this because somewhere in the code we transform `fsub x, 0.0` => `fadd x, 0.0`?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -8100,6 +8100,13 @@ Instruction *InstCombinerImpl::visitFCmpInst(FCmpInst &I) {
         return new FCmpInst(I.getSwappedPredicate(), X, NegC, "", &I);
   }
 
+  // fcmp (fadd X, 0.0), Y --> fcmp X, Y
+  if (match(Op0, m_FAdd(m_Value(X), m_APFloat(C))) && C->isZero()) {
+    if (match(Op1, m_Value(Y))) {
+      return new FCmpInst(Pred, X, Y, "", &I);
+    }
+  }
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8108]:
```suggestion
  if (match(Op0, m_FAdd(m_Value(X), m_AnyZeroFP())))
      return new FCmpInst(Pred, X, Op1, "", &I);
  if (match(Op1, m_FAdd(m_Value(X), m_AnyZeroFP())))
      return new FCmpInst(Pred, Op0, X, "", &I);
```

1. Add a test for `fcmp X, (fadd Y, 0.0)`
2. Add a vector test for `fcmp (fadd X, <0.0, -0.0>), Y`
3. Add a test for fast-math flag preservation

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 1 above)
Comment (vmishelcs) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8108]: > 1. Add a test for `fcmp X, (fadd Y, 0.0)`
> 
>     2. Add a vector test for `fcmp (fadd X, <0.0, -0.0>), Y`
> 
>     3. Add a test for fast-math flag preservation

Should I write a separate test case for each type of `fcmp` instruction (`ugt`, `ogt`, etc) for each of these? Or is one test case sufficient for each?

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8108]: Just one test each is fine.

File: llvm/test/Transforms/InstCombine/fcmp.ll [hunk 2]
```diff
@@ -1284,3 +1284,175 @@ define <1 x i1> @bitcast_1vec_eq0(i32 %x) {
...
+; CHECK-LABEL: @fcmp_fadd_zero_vec(
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp ugt <2 x float> [[X:%.*]], [[Y:%.*]]
+; CHECK-NEXT:    ret <2 x i1> [[CMP]]
+;
+  %add = fadd <2 x float> %x, <float 0.0, float -0.0>
+  %cmp = fcmp ugt <2 x float> %add, %y
+  ret <2 x i1> %cmp
+}
+
+define i1 @fcmp_fadd_zero_fast(float %x, float %y) {
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:1450]: Specifically you have to make sure the flags don't come from the fadd, they cannot be preserved

[the remark this answers is not in the document]
Comment (vmishelcs) [reply llvm/test/Transforms/InstCombine/fcmp.ll]: > Specifically you have to make sure the flags don't come from the fadd, they cannot be preserved

Added! The behaviour seems to be correct from what I can tell.

I have a quick question: How come for some of the tests, the `CHECK` lines generate as
```
define i1 @fcmp_fadd_zero_ugt(float %x, float %y) {
; CHECK-LABEL: @fcmp_fadd_zero_ugt(
; CHECK-NEXT:    [[CMP:%.*]] = fcmp ugt float [[ADD:%.*]], [[Y:%.*]]
; CHECK-NEXT:    ret i1 [[CMP]]
;
  %add = fadd float %x, 0.000000e+00
  %cmp = fcmp ugt float %add, %y
  ret i1 %cmp
}
```
and for the tests which I added later, the `CHECK` lines are
```
define i1 @fcmp_fadd_neg_zero(float %x, float %y) {
; CHECK-LABEL: @fcmp_fadd_neg_zero(
; CHECK-NEXT:    [[CMP:%.*]] = fcmp ugt float [[X:%.*]], [[Y:%.*]]
; CHECK-NEXT:    ret i1 [[CMP]]
;
  %add = fadd float %x, -0.000000e+00
  %cmp = fcmp ugt float %add, %y
  ret i1 %cmp
}
```
?
In particular, why is it that for the former test, the `CHECK` line uses `[[ADD:%.*]]` but the latter uses `[[X:%.*]]`? When I ran the `opt` binary, the result is correct in both cases, but I was wondering why the slight difference.

**Review by arsenm — APPROVED** [review]

**Comment by vmishelcs** [comment]
@dtcxzyw I see there are some issues with the formatting. Is there a script I can run and commit the changes with fixed formatting?

**Comment by dtcxzyw** [comment]
> @dtcxzyw I see there are some issues with the formatting. Is there a script I can run and commit the changes with fixed formatting?

`git clang-format HEAD~1`

**Review by dtcxzyw — APPROVED** [review]
LGTM with the formatting fixed.

**Comment by vmishelcs** [comment]
Formatting should be fixed now.

File: llvm/test/Transforms/InstCombine/fcmp.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/test/Transforms/InstCombine/fcmp.ll]: Sounds like a regression from preserving variable names in UTC, cc @nhaehnle.

**Pull request merged**